core/src/main/java/com/linecorp/armeria/internal/common/DefaultSplitHttpRequest.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.SplitHttpRequest;
+import com.linecorp.armeria.common.stream.SubscriptionOption;
+
+import io.netty.util.concurrent.EventExecutor;
+
+public class DefaultSplitHttpRequest extends AbstractSplitHttpMessage implements SplitHttpRequest {
+
+    private static final HttpHeaders EMPTY_TRAILERS = HttpHeaders.of();
+
+    private final BodySubscriber bodySubscriber = new SplitHttpRequestBodySubscriber();
+    private final HttpRequest request;
+
+    public DefaultSplitHttpRequest(HttpRequest request, EventExecutor executor) {
+        super(request, executor);
+        this.request = request;
+        request.subscribe(bodySubscriber, upstreamExecutor, SubscriptionOption.values());
+    }
+
+    @Override
+    public RequestHeaders headers() {
+        return request.headers();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        subscribe0(subscriber, bodySubscriber, executor, options);
+    }
+
+    private final class SplitHttpRequestBodySubscriber extends BodySubscriber {
+
+        @Override
+        protected void request0(long n) {
+            final Subscription upstream = this.upstream;
```

Comment:
We check if the `Subscription` is null in `DefaultSplitHttpResponse` and I think we should do the same thing here.
How about removing this method and just implementing it in `BodySubscriber`?
We can add `pendingRequests` to `BodySubscriber` as a field and just do:
```java
// In BodySubscriber
@Override
public void onSubscribe(Subscription subscription) {
    super.onSubscribe(subscription);
    if (pendingRequests > 0) {
        subscription.request(pendingRequests);
    }
}
```

Reply:
@minwoox Thank you for the good comment! I fixed it [here!](https://github.com/line/armeria/pull/3953/commits/a76e1104dbe6bb93dd712c76b9516db22770ce49)
If there is any part that needs correction, please feel free to tell me! :-)